src/eddymotion/model/dipy.py

```diff
@@ -0,0 +1,417 @@
+        Exponential covariance function values.
+
+    """
+
+    _ensure_positive_scale(a)
+
+    return np.exp(-theta / a)
+
+
+def compute_spherical_covariance(
+    theta: np.ndarray,
+    a: float,
+) -> np.ndarray:
+    r"""Compute the spherical covariance matrix following eq. 10 in [Andersson15]_.
+
+    .. math::
+
+        C(\theta) = \begin{cases}
+            1 - \frac{3 \theta}{2 a} + \frac{\theta^3}{2 a^3} & \textnormal{if} \; \theta \leq a \\
+            0 & \textnormal{if} \; \theta > a
+        \end{cases}
+
+    Parameters
+    ----------
+    theta : :obj:`~numpy.ndarray`
+        Pairwise angles across diffusion gradient encoding directions.
+    a : :obj:`float`
+        Positive scale parameter that here determines the "distance" at which θ
+        the covariance goes to zero.
+
+    Returns
+    -------
+    :obj:`~numpy.ndarray`
+        Spherical covariance matrix.
+
+    """
+
+    _ensure_positive_scale(a)
+
+    return np.where(theta <= a, 1 - 3 * (theta / a) ** 2 + 2 * (theta / a) ** 3, 0)
+
+
+def compute_pairwise_angles(
+    gtab_X: GradientTable,
+    gtab_Y: GradientTable | None = None,
+    closest_polarity: bool = True,
+) -> np.ndarray:
+    r"""Compute pairwise angles across diffusion gradient encoding directions.
+
+    Following [Andersson15]_, it computes the smallest of the angles between
+    each pair if ``closest_polarity`` is ``True``, i.e.
+
+    .. math::
+
+        \theta(\mathbf{g}, \mathbf{g'}) = \arccos(\abs{\langle \mathbf{g}, \mathbf{g'} \rangle})
+
+    Parameters
+    ----------
+    gtab_X: :obj:`~dipy.core.gradients.GradientTable`
+        Gradient table
+    gtab_Y: :obj:`~dipy.core.gradients.GradientTable`
+        Gradient table
+    closest_polarity : :obj:`bool`
+        ``True`` to consider the smallest of the two angles between the crossing
+         lines resulting from reversing each vector pair.
+
+    Returns
+    -------
+    :obj:`~numpy.ndarray`
+        Pairwise angles across diffusion gradient encoding directions.
+
+    Examples
+    --------
+    >>> compute_pairwise_angles(
+    ...     ((1.0, -1.0), (0.0, 0.0), (0.0, 0.0)),
+    ...     closest_polarity=False,
+    ... )[0, 1]  # doctest: +ELLIPSIS
+    3.1415...
+    >>> compute_pairwise_angles(
+    ...     ((1.0, -1.0), (0.0, 0.0), (0.0, 0.0)),
+    ...     ((1.0, -1.0), (0.0, 0.0), (0.0, 0.0)),
+    ...     closest_polarity=False,
+    ... )[0, 1]  # doctest: +ELLIPSIS
+    3.1415...
+    >>> compute_pairwise_angles(
+    ...     ((1.0, -1.0), (0.0, 0.0), (0.0, 0.0)),
+    ... )[0, 1]
+    0.0
+
+    References
+    ----------
+    .. [Andersson15] J. L. R. Andersson. et al., An integrated approach to
+       correction for off-resonance effects and subject movement in diffusion MR
+       imaging, NeuroImage 125 (2016) 1063–1078
+
+    """
+
+    bvecs_X = gtab_X.bvecs.T
+    bvecs_X = np.array(bvecs_X) / np.linalg.norm(bvecs_X, axis=0)
+
+    if gtab_Y is None:
+        bvecs_Y = bvecs_X
```

Comment:
Support the old use case (auto-covariance) if the second vector is not provided.

Reply:
:+1: Resolving. I'll add a test case for this in a separate PR.